File: client/src/editor/contentid.rs
use viletech::{data::gfx::PictureReader, vfs::FileRef};

/// The editor's best guess at the content in a file.
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub(super) enum ContentId {
	#[default]
	Unknown,
	PlainText,

	/// Action Code Script text source. See <https://doomwiki.org/wiki/ACS>.
	Acs,
	/// Compiled Action Code Script bytecode blob. See <https://doomwiki.org/wiki/ACS>.
	AcsObject,
	/// See <https://zdoom.org/wiki/ALTHUDCF>.
	AltHudCf,
	/// See <https://doomwiki.org/wiki/ANIMATED>.
	Animated,
	/// See <https://zdoom.org/wiki/ANIMDEFS>.
	AnimDefs,
	/// See, for example, <https://doomwiki.org/wiki/ENDOOM>.
	AnsiText,
	/// See <https://doomwiki.org/wiki/Behavior>.
	Behavior,
	/// See <https://doomwiki.org/wiki/Blockmap>.
	Blockmap,
	/// See <https://doomwiki.org/wiki/COLORMAP>.
	Colormap,
	/// See <https://zdoom.org/wiki/CVARINFO>.
	CVarInfo,
	/// See <https://zdoom.org/wiki/DECALDEF>.
	DecalDef,
	/// See <https://zdoom.org/wiki/DECORATE>.
	Decorate,
	/// See <https://zdoom.org/wiki/DEFBINDS>.
	DefBinds,
	/// See <https://zdoom.org/wiki/DEFCVARS>.
	DefCVars,
	/// See <https://doomwiki.org/wiki/DeHackEd>.
	DeHackEd,
	/// See [`Self::DeHackEd`].
	DeHackEdBex,
	/// See <https://doomwiki.org/wiki/Demo>.
	Demo,
	/// See <https://doomwiki.org/wiki/DMXGUS>.
	DmxGus,
	/// See <https://doomwiki.org/wiki/MUS>.
	DmxMus,
	/// See <https://eternity.youfailit.net/wiki/EDF>.
	Edf,
	/// See <https://eternity.youfailit.net/wiki/EMAPINFO>.
	EMapInfo,
	/// See <https://doomwiki.org/wiki/Flat>.
	Flat,
	/// See <https://zdoom.org/wiki/FONTDEFS>.
	FontDefs,
	/// Global FraggleScript. See <https://zdoom.org/wiki/FraggleScript>.
	FsGlobal,
	/// See <https://zdoom.org/wiki/GAMEINFO>.
	GameInfo,
	/// See <https://doomwiki.org/wiki/GENMIDI>.
	GenMidi,
	/// See <https://zdoom.org/wiki/GLDEFS>.
	GlDefs,
	/// See <https://en.wikipedia.org/wiki/OpenGL_Shading_Language>.
	Glsl,
	/// See <https://en.wikipedia.org/wiki/High-Level_Shader_Language>.
	Hlsl,
	/// See <https://zdoom.org/wiki/IWADINFO>.
	IWadInfo,
	/// See <https://zdoom.org/wiki/KEYCONF>.
	KeyConf,
	/// See <https://doomwiki.org/wiki/Linedef>.
	LineDefs,
	/// A [Lithica](viletech::lith) source file.
	Lith,
	/// See <https://zdoom.org/wiki/LOADACS>.
	LoadAcs,
	/// See <https://zdoom.org/wiki/LOCKDEFS>.
	LockDefs,
	/// See <https://doomwiki.org/wiki/MAPINFO>.
	MapInfo,
	/// See <https://doomwiki.org/wiki/WAD#Map_data_lumps>.
	MapMarker,
	/// See <https://en.wikipedia.org/wiki/Markdown>.
	Markdown,
	/// See <https://doomwiki.org/wiki/WAD#Lump_order>.
	Marker,
	/// See <https://doomwiki.org/wiki/MIDI>.
	Midi,
	/// See <https://zdoom.org/wiki/MODELDEF>.
	ModelDef,
	/// See <https://zdoom.org/wiki/MUSINFO>.
	MusInfo,
	/// See <https://doomwiki.org/wiki/Node>.
	Nodes,
	/// See <https://zdoom.org/wiki/PALVERS>.
	PalVers,
	/// See <https://doomwiki.org/wiki/Picture_format>.
	Picture,
	/// A VileTech package metadata text file.
	PkgMeta,
	/// See <https://doomwiki.org/wiki/PLAYPAL>.
	PlayPal,
	/// See <https://doomwiki.org/wiki/PNAMES>.
	PNames,
	/// See <https://en.wikipedia.org/wiki/PNG>.
	Png,
	/// See <https://doomwiki.org/wiki/Reject>.
	Reject,
	/// See <https://zdoom.org/wiki/SBARINFO>.
	SBarInfo,
	/// See <https://doomwiki.org/wiki/Sector>.
	Sectors,
	/// See <https://doomwiki.org/wiki/Seg>.
	Segs,
	/// See <https://doomwiki.org/wiki/Sidedef>.
	SideDefs,
	/// See <https://zdoom.org/wiki/SNDINFO>.
	SndInfo,
	/// See <https://zdoom.org/wiki/SNDSEQ>.
	SndSeq,
	/// See <https://doomwiki.org/wiki/Subsector>.
	SSectors,
	/// See <https://doomwiki.org/wiki/SWITCHES>.
	Switches,
	/// See <https://zdoom.org/wiki/TERRAIN>.
	Terrain,
	/// See <https://zdoom.org/wiki/TEXTCOLO>.
	TextColo,
	/// UDMF TEXTMAP lump. See <https://doomwiki.org/wiki/UDMF>.
	TextMap,
	/// See <https://doomwiki.org/wiki/TEXTURE1_and_TEXTURE2>.
	TextureX,
	/// See <https://doomwiki.org/wiki/Thing>.
	Things,
	/// See <https://zdoom.org/wiki/TRNSLATE>.
	Trnslate,
	/// See <https://doomwiki.org/wiki/UMAPINFO>.
	UMapInfo,
	/// See <https://doomwiki.org/wiki/Vertex>.
	Vertexes,
	/// See <https://zdoom.org/wiki/VOXELDEF>.
	VoxelDef,
	/// An informal standard for level pack metadata.
	WadInfo,
	/// See <https://www.w3.org/TR/WGSL/>.
	Wgsl,
	/// See <https://zdoom.org/wiki/X11R6RGB>.
	X11R6RGB,
	/// See <https://zdoom.org/wiki/XHAIRS>.
	XHairs,
	/// [`Self::MapInfo`], but ZDoom-specific.
	ZMapInfo,
	/// See <https://zdoom.org/wiki/ZScript>.
	ZScript,
}

impl std::fmt::Display for ContentId {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::Unknown => write!(f, "Unknown"),
			Self::PlainText => write!(f, "Plain text"),

			Self::Acs => write!(f, "ACS source"),
			Self::AcsObject => write!(f, "Compiled ACS"),
			Self::AltHudCf => write!(f, "ZDoom alt. HUD config."),
			Self::Animated => write!(f, "Boom animations"),
			Self::AnimDefs => write!(f, "ZDoom animations"),
			Self::AnsiText => write!(f, "ANSI text"),
			Self::Behavior => write!(f, "ACS map scripts"),
			Self::Blockmap => write!(f, "Blockmap data"),
			Self::Colormap => write!(f, "Colormap"),
			Self::CVarInfo => write!(f, "CVar defs"),
			Self::DecalDef => write!(f, "ZDoom decal config."),
			Self::Decorate => write!(f, "DECORATE"),
			Self::DefBinds => write!(f, "ZDoom keybind defaults"),
			Self::DefCVars => write!(f, "ZDoom CVar defaults"),
			Self::DeHackEd => write!(f, "DeHackEd"),
			Self::DeHackEdBex => write!(f, "DeH. Boom EXtended"),
			Self::Demo => write!(f, "Demo"),
			Self::DmxMus => write!(f, "MUS audio"),
			Self::DmxGus => write!(f, "DMX GUS config."),
			Self::Edf => write!(f, "Eternity definitions"),
			Self::EMapInfo => write!(f, "Eternity map info."),
			Self::Flat => write!(f, "Graphic (Flat)"),
			Self::FontDefs => write!(f, "ZDoom font config."),
			Self::FsGlobal => write!(f, "Global FraggleScript"),
			Self::GameInfo => write!(f, "ZDoom game config."),
			Self::GenMidi => write!(f, "General MIDI set"),
			Self::GlDefs => write!(f, "OpenGL config."),
			Self::Glsl => write!(f, "OpenGL shader"),
			Self::Hlsl => write!(f, "Direct3D shader"),
			Self::IWadInfo => write!(f, "ZDoom IWAD config."),
			Self::KeyConf => write!(f, "ZDoom keybindings"),
			Self::Lith => write!(f, "Lithica source"),
			Self::LineDefs => write!(f, "Map lines"),
			Self::LoadAcs => write!(f, "ACS loader"),
			Self::LockDefs => write!(f, "ZDoom lock/key config."),
			Self::MapInfo => write!(f, "Map info"),
			Self::MapMarker => write!(f, "Map marker"),
			Self::Markdown => write!(f, "Markdown"),
			Self::Marker => write!(f, "Marker"),
			Self::Midi => write!(f, "MIDI audio"),
			Self::MusInfo => write!(f, "ZDoom music config."),
			Self::ModelDef => write!(f, "ZDoom model config."),
			Self::Nodes => write!(f, "Map BSP nodes"),
			Self::PalVers => write!(f, "ZDoom palette versioning"),
			Self::Picture => write!(f, "Graphic (picture)"),
			Self::PkgMeta => write!(f, "VileTech package metadata"),
			Self::PlayPal => write!(f, "Color palette"),
			Self::PNames => write!(f, "Patch table"),
			Self::Png => write!(f, "Graphic (PNG)"),
			Self::Reject => write!(f, "Map reject table"),
			Self::SBarInfo => write!(f, "ZDoom HUD config."),
			Self::Sectors => write!(f, "Map sectors"),
			Self::Segs => write!(f, "Map segments"),
			Self::SideDefs => write!(f, "Map sides"),
			Self::SndInfo => write!(f, "ZDoom sound config."),
			Self::SndSeq => write!(f, "ZDoom sound sequence"),
			Self::SSectors => write!(f, "Map subsectors"),
			Self::Switches => write!(f, "Boom switches"),
			Self::Terrain => write!(f, "ZDoom terrain config."),
			Self::TextColo => write!(f, "Font colors"),
			Self::TextMap => write!(f, "UDMF textmap"),
			Self::TextureX => write!(f, "TEXTURE1/TEXTURE2"),
			Self::Things => write!(f, "Map object placements"),
			Self::Trnslate => write!(f, "ZDoom translations"),
			Self::UMapInfo => write!(f, "Universal map info"),
			Self::Vertexes => write!(f, "Map vertices"),
			Self::VoxelDef => write!(f, "ZDoom voxel config."),
			Self::WadInfo => write!(f, "WAD metadata"),
			Self::Wgsl => write!(f, "WebGPU shader"),
			Self::X11R6RGB => write!(f, "ZDoom X11R6RGB"),
			Self::XHairs => write!(f, "ZDoom crosshair config."),
			Self::ZMapInfo => write!(f, "ZDoom map info"),
			Self::ZScript => write!(f, "ZScript"),
		}
	}
}

impl ContentId {
	/// TODO: might be worthwhile to make this a hashmap.
	const FILE_STEMS: &'static [(&'static str, Self)] = &[
		("ALTHUDCF", Self::AltHudCf),
		("ANIMATED", Self::Animated),
		("ANIMDEFS", Self::AnimDefs),
		("BEHAVIOR", Self::Behavior),
		("BLOCKMAP", Self::Blockmap),
		("COLORMAP", Self::Colormap),
		("CVARINFO", Self::CVarInfo),
		("DECALDEF", Self::DecalDef),
		("DECORATE", Self::Decorate),
		("DEFBINDS", Self::DefBinds),
		("DEFCVARS", Self::DefCVars),
		("DEHACKED", Self::DeHackEd),
		("DMXGUS", Self::DmxGus),
		("DMXGUSC", Self::DmxGus),
		("EMAPINFO", Self::EMapInfo),
		("ENDOOM", Self::AnsiText),
		("FONTDEFS", Self::FontDefs),
		("FSGLOBAL", Self::FsGlobal),
		("GAMEINFO", Self::GameInfo),
		("GENMIDI", Self::GenMidi),
		("GLDEFS", Self::GlDefs),
		("IWADINFO", Self::IWadInfo),
		("KEYCONF", Self::KeyConf),
		("LINEDEFS", Self::LineDefs),
		("LOADACS", Self::LoadAcs),
		("LOCKDEFS", Self::LockDefs),
		("MAPINFO", Self::MapInfo),
		("MODELDEF", Self::ModelDef),
		("MUSINFO", Self::MusInfo),
		("NODES", Self::Nodes),
		("PALVERS", Self::PalVers),
		("PLAYPAL", Self::PlayPal),
		("PNAMES", Self::PNames),
		("REJECT", Self::Reject),
		("SBARINFO", Self::SBarInfo),
		("SECTORS", Self::Sectors),
		("SEGS", Self::Segs),
		("SIDEDEFS", Self::SideDefs),
		("SNDINFO", Self::SndInfo),
		("SNDSEQ", Self::SndSeq),
		("SSECTORS", Self::SSectors),
		("SWITCHES", Self::Switches),
		("TERRAIN", Self::Terrain),
		("TEXTCOLO", Self::TextColo),
		("TEXTMAP", Self::TextMap),
		("TEXTURE1", Self::TextureX),
		("TEXTURE2", Self::TextureX),
		("THINGS", Self::Things),
		("TRNSLATE", Self::Trnslate),
		("UMAPINFO", Self::UMapInfo),
		("VERTEXES", Self::Vertexes),
		("VOXELDEF", Self::VoxelDef),
		("WADINFO", Self::WadInfo),
		("X11R6RGB", Self::X11R6RGB),
		("XHAIRS", Self::XHairs),
		("ZMAPINFO", Self::ZMapInfo),
		("ZSCRIPT", Self::ZScript),
	];

	const FILE_STEM_MARKER_PREFIXES: &'static [&'static str] = &[
		"A_", "F_", "F1_", "F2_", "HI_", "P_", "PP_", "P1_", "P2_", "P3_", "S_", "SS_", "T_", "TX_",
	];

	const EXTENSIONS: &'static [(&'static str, Self)] = &[
		("acs", Self::Acs),
		("bex", Self::DeHackEdBex),
		("dec", Self::Decorate),
		("deh", Self::DeHackEd),
		("edf", Self::Edf),
		("glsl", Self::Glsl),
		("hlsl", Self::Hlsl),
		("lith", Self::Lith),
		("md", Self::Markdown),
		("txt", Self::PlainText),
		("wgsl", Self::Wgsl),
		("zs", Self::ZScript),
		("zsc", Self::ZScript),
	];

	#[must_use]
	pub(super) fn is_text(self) -> bool {
		// Don't change this to use `matches`.
		// We want a compiler error to get raised if new, unhandled variants are added.

		match self {
			Self::PlainText
			| Self::Acs
			| Self::AltHudCf
			| Self::CVarInfo
			| Self::DecalDef
			| Self::Decorate
			| Self::DefBinds
			| Self::DefCVars
			| Self::DeHackEd
			| Self::DeHackEdBex
			| Self::Edf
			| Self::EMapInfo
			| Self::FontDefs
			| Self::FsGlobal
			| Self::GameInfo
			| Self::GenMidi
			| Self::GlDefs
			| Self::Glsl
			| Self::Hlsl
			| Self::IWadInfo
			| Self::KeyConf
			| Self::Lith
			| Self::LoadAcs
			| Self::LockDefs
			| Self::MapInfo
			| Self::Markdown
			| Self::ModelDef
			| Self::MusInfo
			| Self::PalVers
			| Self::PkgMeta
			| Self::SBarInfo
			| Self::SndInfo
			| Self::Terrain
			| Self::TextColo
			| Self::TextMap
			| Self::Trnslate
			| Self::UMapInfo
			| Self::VoxelDef
			| Self::WadInfo
			| Self::Wgsl
			| Self::X11R6RGB
			| Self::XHairs
			| Self::ZMapInfo
			| Self::ZScript => true,
			Self::AnimDefs
			| Self::Unknown
			| Self::AcsObject
			| Self::Animated
			| Self::AnsiText
			| Self::Behavior
			| Self::Blockmap
			| Self::Colormap
			| Self::Demo
			| Self::DmxGus
			| Self::DmxMus
			| Self::Flat
			| Self::LineDefs
			| Self::MapMarker
			| Self::Marker
			| Self::Midi
			| Self::Nodes
			| Self::Picture
			| Self::PlayPal
			| Self::PNames
			| Self::Png
			| Self::Reject
			| Self::Sectors
			| Self::Segs
			| Self::SideDefs
			| Self::SndSeq
			| Self::SSectors
			| Self::Switches
			| Self::TextureX
			| Self::Things
			| Self::Vertexes => false,
		}
	}

	#[must_use]
	pub(super) fn deduce(vfile: &FileRef, bytes: &[u8], markers: WadMarkers) -> Self {
		if let Some(next) = vfile.next_sibling() {
			if next.name().eq_ignore_ascii_case("THINGS")
				|| next.name().eq_ignore_ascii_case("TEXTMAP")
			{
				return Self::MapMarker;
			}
		}

		let fname = vfile.name();

		if let Some(ext) = fname.extension() {
			if fname
				.file_prefix()
				.is_some_and(|p| p.eq_ignore_ascii_case("meta"))
				&& ext.eq_ignore_ascii_case("toml")
			{
				return Self::PkgMeta;
			}

			if let Some(tuple) = Self::EXTENSIONS
				.iter()
				.find(|t| t.0.eq_ignore_ascii_case(ext))
			{
				return tuple.1;
			}
		}

		if let Some(fstem) = fname.file_stem() {
			let fstem = fstem.as_str();

			if bytes.is_empty() {
				let buf = fstem.to_ascii_uppercase();

				if Self::FILE_STEM_MARKER_PREFIXES
					.iter()
					.any(|&p| buf.starts_with(p))
					&& (buf.ends_with("START") || buf.ends_with("END"))
				{
					return ContentId::Marker;
				}
			}

			if let Some(tuple) = Self::FILE_STEMS
				.iter()
				.find(|t| t.0.eq_ignore_ascii_case(fstem))
			{
				return tuple.1;
			}

			if fstem
				.get(0..4)
				.is_some_and(|f4| f4.eq_ignore_ascii_case("demo"))
			{
				return Self::Demo;
			}
		}

		if viletech::mus::is_dmxmus(bytes) {
			return Self::DmxMus;
		}

		if viletech::audio::MidiFormat::deduce(bytes).is_some() {
			return Self::Midi;
		}

		if viletech::util::io::is_png(bytes) {
			return Self::Png;
		}

		if viletech::data::acs::is_object(bytes) {
			return Self::AcsObject;
		}

		match markers {
			WadMarkers::None => {}
			WadMarkers::Flats => {
				if bytes.len() == 4096 {
					return Self::Flat;
				}
			}
		}

		if PictureReader::new(bytes).is_ok() {
			return Self::Picture;
		}

		Self::Unknown
	}
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum WadMarkers {
	None,
	Flats,
}

File: engine/src/contentid.zig
// Any code herein taken from the "infer" crate is used under the MIT License.
// See `/legal/infer.txt`.

const std = @import("std");

/// A tag for a file or archive entry representing the result of heuristics used
/// to determine what its content is most likely to be.
pub const ContentId = enum {
    unknown,
    plain_text,

    /// Action Code Script text source. See <https://doomwiki.org/wiki/ACS>.
    acs,
    /// Compiled Action Code Script bytecode blob. See <https://doomwiki.org/wiki/ACS>.
    acs_object,
    /// See <https://zdoom.org/wiki/ALTHUDCF>.
    althudcf,
    /// See <https://doomwiki.org/wiki/ANIMATED>.
    animated,
    /// See <https://zdoom.org/wiki/ANIMDEFS>.
    animdefs,
    /// See, for example, <https://doomwiki.org/wiki/ENDOOM>.
    ansi_text,
    /// See <https://doomwiki.org/wiki/Behavior>.
    behavior,
    /// See <https://doomwiki.org/wiki/Blockmap>.
    blockmap,
    /// See <https://doomwiki.org/wiki/COLORMAP>.
    colormap,
    /// See <https://doomwiki.org/wiki/COMPLVL>.
    complvl,
    /// See <https://zdoom.org/wiki/CVARINFO>.
    cvarinfo,
    /// See <https://zdoom.org/wiki/DECALDEF>.
    decaldef,
    /// See <https://mtrop.github.io/DoomTools/decohack.html>.
    decohack,
    /// See <https://zdoom.org/wiki/DECORATE>.
    decorate,
    /// See <https://zdoom.org/wiki/DEFBINDS>.
    defbinds,
    /// See <https://zdoom.org/wiki/DEFCVARS>.
    defcvars,
    /// See <https://doomwiki.org/wiki/DeHackEd>.
    dehacked,
    /// See [`dehacked`].
    dehacked_bex,
    /// See <https://doomwiki.org/wiki/Demo>.
    demo,
    /// See <https://doomwiki.org/wiki/DEMOLOOP>.
    demoloop,
    /// See <https://doomwiki.org/wiki/DMXGUS>.
    dmxgus,
    /// See <https://doomwiki.org/wiki/MUS>.
    dmx_mus,
    /// Eternity Definition File. See <https://eternity.youfailit.net/wiki/EDF>.
    edf,
    /// Eternity Map Information. See <https://eternity.youfailit.net/wiki/EMAPINFO>.
    emapinfo,
    /// See <https://en.wikipedia.org/wiki/FLAC>.
    flac,
    /// A 64-by-64 texture for floors and ceilings. See <https://doomwiki.org/wiki/Flat>.
    flat,
    /// See <https://zdoom.org/wiki/FONTDEFS>.
    fontdefs,
    /// Global FraggleScript. See <https://zdoom.org/wiki/FraggleScript>.
    fsglobal,
    /// See <https://doomwiki.org/wiki/GAMECONF>.
    gameconf,
    /// See <https://zdoom.org/wiki/GAMEINFO>.
    gameinfo,
    /// See <https://doomwiki.org/wiki/GENMIDI>.
    genmidi,
    /// See <https://zdoom.org/wiki/GLDEFS>.
    gldefs,
    /// See <https://en.wikipedia.org/wiki/OpenGL_Shading_Language>.
    glsl,
    /// See <https://en.wikipedia.org/wiki/High-Level_Shader_Language>.
    hlsl,
    /// See <https://zdoom.org/wiki/IWADINFO>.
    iwadinfo,
    /// See <https://en.wikipedia.org/wiki/JavaScript>.
    javascript, // Likely UDBScript...
    /// See <https://en.wikipedia.org/wiki/JPEG>.
    jpeg,
    /// See <https://en.wikipedia.org/wiki/JPEG_XL>.
    jpeg_xl,
    /// See <https://en.wikipedia.org/wiki/JSON>.
    json,
    /// See <https://zdoom.org/wiki/KEYCONF>.
    keyconf,
    /// See <https://doomwiki.org/wiki/Linedef>.
    linedefs,
    /// See <https://zdoom.org/wiki/LOADACS>.
    loadacs,
    /// See <https://zdoom.org/wiki/LOCKDEFS>.
    lockdefs,
    /// See <https://doomwiki.org/wiki/MAPINFO>.
    mapinfo,
    /// See <https://doomwiki.org/wiki/WAD#Map_data_lumps>.
    map_marker,
    /// See <https://en.wikipedia.org/wiki/Markdown>.
    markdown,
    /// See <https://doomwiki.org/wiki/WAD#Lump_order>.
    marker,
    /// See <https://doomwiki.org/wiki/MIDI>.
    midi,
    /// See <https://en.wikipedia.org/wiki/MP3>.
    mp3,
    /// See <https://zdoom.org/wiki/MODELDEF>.
    modeldef,
    /// See <https://doomwiki.org/wiki/MUSINFO>.
    musinfo,
    /// See <https://doomwiki.org/wiki/Node>.
    nodes,
    /// See <https://zdoom.org/wiki/PALVERS>.
    palvers,
    /// See <https://doomwiki.org/wiki/Picture_format>.
    picture,
    /// See <https://doomwiki.org/wiki/PLAYPAL>.
    playpal,
    /// See <https://doomwiki.org/wiki/PNAMES>.
    pnames,
    /// See <https://en.wikipedia.org/wiki/PNG>.
    png,
    /// See <https://doomwiki.org/wiki/Reject>.
    reject,
    /// See <https://doomwiki.org/wiki/SBARDEF>.
    sbardef,
    /// See <https://zdoom.org/wiki/SBARINFO>.
    sbarinfo,
    /// See <https://doomwiki.org/wiki/Sector>.
    sectors,
    /// See <https://doomwiki.org/wiki/Seg>.
    segs,
    /// See <https://doomwiki.org/wiki/Sidedef>.
    sidedefs,
    /// See <https://doomwiki.org/wiki/SKYDEFS>.
    skydefs,
    /// See <https://zdoom.org/wiki/SNDINFO>.
    sndinfo,
    /// See <https://zdoom.org/wiki/SNDSEQ>.
    sndseq,
    /// See <https://doomwiki.org/wiki/Subsector>.
    ssectors,
    /// See <https://en.wikipedia.org/wiki/SVG>.
    svg,
    /// See <https://doomwiki.org/wiki/SWITCHES>.
    switches,
    /// See <https://zdoom.org/wiki/TERRAIN>.
    terrain,
    /// See <https://zdoom.org/wiki/TEXTCOLO>.
    textcolo,
    /// UDMF TEXTMAP lump. See <https://doomwiki.org/wiki/UDMF>.
    textmap,
    /// See <https://doomwiki.org/wiki/TEXTURE1_and_TEXTURE2>.
    texturex,
    /// See <https://doomwiki.org/wiki/Thing>.
    things,
    /// See <https://zdoom.org/wiki/TRNSLATE>.
    trnslate,
    /// Universal Map Information. See <https://doomwiki.org/wiki/UMAPINFO>.
    umapinfo,
    /// See <https://doomwiki.org/wiki/Vertex>.
    vertexes,
    /// See <https://zdoom.org/wiki/VOXELDEF>.
    voxeldef,
    /// An informal standard for level pack metadata.
    wadinfo,
    /// See <https://en.wikipedia.org/wiki/WebP>.
    webp,
    /// See <https://www.w3.org/TR/WGSL/>.
    wgsl,
    /// See <https://zdoom.org/wiki/X11R6RGB>.
    x11r6rgb,
    /// See <https://zdoom.org/wiki/XHAIRS>.
    xhairs,
    /// [`mapinfo`], but ZDoom-specific.
    zmapinfo,
    /// See <https://zdoom.org/wiki/ZScript>.
    zscript,

    pub const by_filestem = std.StaticStringMapWithEql(
        ContentId,
        std.ascii.eqlIgnoreCase,
    ).initComptime(.{
        .{ "althudcf", .althudcf },
        .{ "animated", .animated },
        .{ "animdefs", .animdefs },
        .{ "behavior", .behavior },
        .{ "blockmap", .blockmap },
        .{ "colormap", .colormap },
        .{ "cvarinfo", .cvarinfo },
        .{ "decaldef", .decaldef },
        .{ "decorate", .decorate },
        .{ "defbinds", .defbinds },
        .{ "defcvars", .defcvars },
        .{ "dehacked", .dehacked },
        .{ "demoloop", .demoloop },
        .{ "dmxgus", .dmxgus },
        .{ "dmxgusc", .dmxgus },
        .{ "emapinfo", .emapinfo },
        .{ "emenus", .edf },
        .{ "estrings", .edf },
        .{ "endoom", .ansi_text },
        .{ "fontdefs", .fontdefs },
        .{ "fsglobal", .fsglobal },
        .{ "gameconf", .gameconf },
        .{ "gameinfo", .gameinfo },
        .{ "genmidi", .genmidi },
        .{ "gldefs", .gldefs },
        .{ "iwadinfo", .iwadinfo },
        .{ "keyconf", .keyconf },
        .{ "linedefs", .linedefs },
        .{ "loadacs", .loadacs },
        .{ "lockdefs", .lockdefs },
        .{ "mapinfo", .mapinfo },
        .{ "modeldef", .modeldef },
        .{ "musinfo", .musinfo },
        .{ "nodes", .nodes },
        .{ "palvers", .palvers },
        .{ "playpal", .playpal },
        .{ "pnames", .pnames },
        .{ "reject", .reject },
        .{ "sbarinfo", .sbarinfo },
        .{ "sectors", .sectors },
        .{ "segs", .segs },
        .{ "sidedefs", .sidedefs },
        .{ "sndinfo", .sndinfo },
        .{ "sndseq", .sndseq },
        .{ "ssectors", .ssectors },
        .{ "switches", .switches },
        .{ "terrain", .terrain },
        .{ "textcolo", .textcolo },
        .{ "textmap", .textmap },
        .{ "texture1", .texturex },
        .{ "texture2", .texturex },
        .{ "things", .things },
        .{ "trnslate", .trnslate },
        .{ "umapinfo", .umapinfo },
        .{ "vertexes", .vertexes },
        .{ "voxeldef", .voxeldef },
        .{ "wadinfo", .wadinfo },
        .{ "x11r6rgb", .x11r6rgb },
        .{ "xhairs", .xhairs },
        .{ "zmapinfo", .zmapinfo },
        .{ "zscript", .zscript },
    });

    pub const by_extension = std.StaticStringMapWithEql(
        ContentId,
        std.ascii.eqlIgnoreCase,
    ).initComptime(.{
        .{ ".acs", .acs },
        .{ ".bex", .dehacked_bex },
        .{ ".dec", .decorate },
        .{ ".deh", .dehacked },
        .{ ".dh", .decohack },
        .{ ".edf", .edf },
        .{ ".flac", .flac },
        .{ ".frag", .glsl },
        .{ ".glsl", .glsl },
        .{ ".hlsl", .hlsl },
        .{ ".jfi", .jpeg },
        .{ ".jfif", .jpeg },
        .{ ".jif", .jpeg },
        .{ ".jpe", .jpeg },
        .{ ".jpeg", .jpeg },
        .{ ".jpg", .jpeg },
        .{ ".js", .javascript },
        .{ ".json", .json },
        .{ ".jxl", .jpeg_xl },
        .{ ".md", .markdown },
        .{ ".mid", .midi },
        .{ ".mp3", .mp3 },
        .{ ".mus", .dmx_mus },
        .{ ".svg", .svg },
        .{ ".txt", .plain_text },
        .{ ".vert", .glsl },
        .{ ".webp", .webp },
        .{ ".wgsl", .wgsl },
        .{ ".zs", .zscript },
        .{ ".zsc", .zscript },
    });

    pub fn prettyName(self: ContentId) []const u8 {
        return switch (self) {
            .unknown => "Unknown",
            .plain_text => "Plain text",

            .acs => "ACS source",
            .acs_object => "ACS bytecode",
            .althudcf => "ZDoom alt. HUD config.",
            .animated => "Boom animations",
            .animdefs => "ZDoom animations",
            .ansi_text => "ANSI text",
            .behavior => "ACS map scripts",
            .blockmap => "Blockmap data",
            .colormap => "Colormap",
            .complvl => "Compatibility hint",
            .cvarinfo => "CVar defs",
            .decaldef => "ZDoom decal config.",
            .decorate => "DECORATE",
            .decohack => "DECOHack",
            .defbinds => "ZDoom keybind defaults",
            .defcvars => "ZDoom CVar defaults",
            .dehacked => "DeHackEd",
            .dehacked_bex => "DeH. Boom EXtended",
            .demo => "Demo",
            .demoloop => "ID24 Demo-loop config.",
            .dmx_mus => "Audio (DMX MUS)",
            .dmxgus => "DMX GUS config.",
            .edf => "Eternity definitions",
            .emapinfo => "Eternity map info.",
            .flac => "Audio (FLAC)",
            .flat => "Graphic (Flat)",
            .fontdefs => "ZDoom font config.",
            .fsglobal => "Global FraggleScript",
            .gameconf => "ID24 game config.",
            .gameinfo => "ZDoom game config.",
            .genmidi => "General MIDI set",
            .gldefs => "OpenGL config.",
            .glsl => "OpenGL shader",
            .hlsl => "Direct3D shader",
            .iwadinfo => "ZDoom IWAD config.",
            .javascript => "JavaScript",
            .jpeg => "Graphic (JPEG)",
            .jpeg_xl => "Graphic (JPEG XL)",
            .json => "JSON",
            .keyconf => "ZDoom keybindings",
            .linedefs => "Map lines",
            .loadacs => "ACS loader",
            .lockdefs => "ZDoom lock/key config.",
            .mapinfo => "Map info",
            .map_marker => "Map marker",
            .markdown => "Markdown",
            .marker => "Marker",
            .midi => "Audio (MIDI)",
            .modeldef => "ZDoom model config.",
            .mp3 => "Audio (MP3)",
            .musinfo => "ZDoom music config.",
            .nodes => "Map BSP nodes",
            .palvers => "ZDoom palette versioning",
            .picture => "Graphic (picture)",
            .playpal => "Color palette",
            .pnames => "Patch table",
            .png => "Graphic (PNG)",
            .reject => "Map reject table",
            .sbardef => "ID24 status bar config.",
            .sbarinfo => "ZDoom HUD config.",
            .sectors => "Map sectors",
            .segs => "Map segments",
            .sidedefs => "Map sides",
            .skydefs => "ID24 sky config.",
            .sndinfo => "ZDoom sound config.",
            .sndseq => "ZDoom sound sequence",
            .ssectors => "Map subsectors",
            .svg => "Graphic (SVG)",
            .switches => "Boom switches",
            .terrain => "ZDoom terrain config.",
            .textcolo => "Font colors",
            .textmap => "UDMF textmap",
            .texturex => "TEXTURE1/TEXTURE2",
            .things => "Map object placements",
            .trnslate => "ZDoom translations",
            .umapinfo => "Universal map info",
            .vertexes => "Map vertices",
            .voxeldef => "ZDoom voxel config.",
            .wadinfo => "WAD metadata",
            .webp => "Graphic (WebP)",
            .wgsl => "WebGPU shader",
            .x11r6rgb => "ZDoom X11R6RGB",
            .xhairs => "ZDoom crosshair config.",
            .zmapinfo => "ZDoom map info",
            .zscript => "ZScript",
        };
    }

    pub fn isFlac(buf: []const u8) bool {
        // https://docs.rs/infer/0.16.0/src/infer/matchers/audio.rs.html#49-52
        return buf.len > 3 and std.mem.eql(u8, buf[0..4], "\x66\x4c\x61\x43");
    }

    pub fn isMp3(buf: []const u8) bool {
        // https://docs.rs/infer/0.16.0/src/infer/matchers/audio.rs.html#7-12
        return buf.len > 2 and ((buf[0] == 0x49 and buf[1] == 0x44 and buf[2] == 0x33) // ID3v2
        // Final bit (has crc32) may be or may not be set.
        || (buf[0] == 0xFF and buf[1] == 0xFB));
    }
};

test "ContentId, semantic check" {
    _ = ContentId.by_filestem.get("");
    _ = ContentId.by_extension.get("");
    _ = ContentId.prettyName(.unknown);
}
